be/src/exec/json_scanner.h

```diff
@@ -0,0 +1,146 @@
+#include "runtime/descriptors.h"
+#include "runtime/stream_load/load_stream_mgr.h"
+#include "runtime/small_file_mgr.h"
+
+namespace doris {
+class Tuple;
+class SlotDescriptor;
+class RuntimeState;
+class TupleDescriptor;
+class MemTracker;
+class JsonReader;
+
+class JsonScanner : public BaseScanner {
+public:
+    JsonScanner(
+        RuntimeState* state,
+        RuntimeProfile* profile,
+        const TBrokerScanRangeParams& params,
+        const std::vector<TBrokerRangeDesc>& ranges,
+        const std::vector<TNetworkAddress>& broker_addresses,
+        ScannerCounter* counter);
+    ~JsonScanner();
+
+    // Open this scanner, will initialize information need to
+    Status open() override;
+
+    // Get next tuple
+    Status get_next(Tuple* tuple, MemPool* tuple_pool, bool* eof) override;
+
+    // Close this scanner
+    void close() override;
+private:
+    Status open_next_reader();
+private:
+    const std::vector<TBrokerRangeDesc>& _ranges;
+    const std::vector<TNetworkAddress>& _broker_addresses;
+
+    std::string _jsonpath;
+    std::string _jsonpath_file;
+    // std::shared_ptr<rapidjson::Document> _jsonPathDoc;
+
+    // used to hold current StreamLoadPipe
+    std::shared_ptr<StreamLoadPipe> _stream_load_pipe;
+    // Reader
+    JsonReader* _cur_file_reader;
+    int _next_range;
+    bool _cur_file_eof; // is read over?
+    bool _scanner_eof;
+};
+
+
+class JsonDataInternal {
+public:
+    JsonDataInternal(rapidjson::Value* v);
+    ~JsonDataInternal();
+    rapidjson::Value::ConstValueIterator get_next();
+
+private:
+    bool is_end();
+
+private:
+    rapidjson::Value* _json_values;
+    rapidjson::Value::ConstValueIterator  _iterator;
+};
+
+class JsonReader {
+public:
+    JsonReader(RuntimeState* state, ScannerCounter* counter, SmallFileMgr *fileMgr, RuntimeProfile* profile, FileReader* file_reader,
+            std::string& jsonpath, std::string& jsonpath_file);
+    ~JsonReader();
+    Status read(Tuple* tuple, std::vector<SlotDescriptor*> slot_descs, MemPool* tuple_pool, bool* eof);
+
+private:
+    void init_jsonpath(SmallFileMgr *fileMgr, std::string& jsonpath, std::string& jsonpath_file);
+    void fill_slot(Tuple* tuple, SlotDescriptor* slot_desc, MemPool* mem_pool, const uint8_t* value, int32_t len);
+    size_t get_data_by_jsonpath();
+    int parse_jsonpath_from_file(SmallFileMgr *smallFileMgr, std::string& fileinfo);
+    Status parse_json_doc(bool *eof);
+    Status set_tuple_value(rapidjson::Value& objectValue, Tuple* tuple, std::vector<SlotDescriptor*> slot_descs, MemPool* tuple_pool, bool *valid, int *nullcount);
+    Status set_tuple_value_from_map(Tuple* tuple, std::vector<SlotDescriptor*> slot_descs, MemPool* tuple_pool, bool *valid);
+    Status handle_simple_json(Tuple* tuple, std::vector<SlotDescriptor*> slot_descs, MemPool* tuple_pool, bool* eof);
+    Status handle_complex_json(Tuple* tuple, std::vector<SlotDescriptor*> slot_descs, MemPool* tuple_pool, bool* eof);
```

Comment:
```suggestion
    Status handle_complex_json(Tuple* tuple, const std::vector<SlotDescriptor*>& slot_descs, MemPool* tuple_pool, bool* eof);
```
There are many `slot_descs` in other places has same problem.